overseer/src/lib.rs

```diff
@@ -0,0 +1,552 @@
+//!             .           |                                       |            .
+//!             ..................................................................
+//!                         |                                       |
+//!                       start()                                 start()
+//!                         V                                       V
+//!             ..................| Overseer "runs" these |.......................
+//!             .  +-------------------+                +---------------------+  .
+//!             .  | SubsystemInstance1|                | SubsystemInstance2  |  .
+//!             .  +-------------------+                +---------------------+  .
+//!             ..................................................................
+//! ```
+
+use std::fmt::Debug;
+use std::pin::Pin;
+use std::collections::{HashSet, HashMap};
+use std::task::Poll;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use futures::channel::{mpsc, oneshot};
+use futures::{
+	pending, poll,
+	Future, SinkExt, StreamExt,
+};
+
+
+// TODO: a better way to get unique IDs for jobs
+fn get_id() -> usize {
+	static COUNTER:AtomicUsize = AtomicUsize::new(1);
+	COUNTER.fetch_add(1, Ordering::Relaxed)
+}
+
+/// An error type that describes faults that may happen
+///
+/// These are:
+///   * Channels being closed
+///   * Subsystems dying when they are not expected to
+///   * Subsystems not dying when they are told to die
+///   * etc.
+// TODO: populate with actual error cases.
+#[derive(Debug)]
+pub struct SubsystemError;
+
+/// A `Result` type that wraps `SubsystemError` and an empty type on success.
+// TODO: Proper success type.
+pub type SubsystemResult = Result<(), SubsystemError>;
+
+/// An asynchronous job that runs inside and being overseen by the `Overseer`.
+///
+/// In essence it's just a newtype wrapping a pinned `Future` dyn trait object.
+pub struct SubsystemJob(pub Pin<Box<dyn Future<Output = SubsystemResult>>>);
+
+/// A type of messages that are used inside the `Overseer`.
+///
+/// It is generic over some `T` that is intended to be a message type
+/// being used by the subsystems running on the `Overseer`. Most likely
+/// this type will be one large `enum` covering all possible messages in
+/// the system.
+enum OverseerMessage<T: Debug> {
+	/// This is a message generated by a `Subsystem`.
+	/// Wraps the messge itself and has an optional `to` of
+	/// someone who can receive this message.
+	///
+	/// If that `to` is present the message will be targetedly sent to the intended
+	/// receiver. The most obvious use case of this is communicating with children.
+	SubsystemMessage{
+		to: Option<SubsystemId>,
+		msg: T,
+	},
+	/// A message that wraps something the `Subsystem` is desiring to
+	/// spawn on the overseer and a `oneshot::Sender` to signal the result
+	/// of the spawn.
+	Spawn{
+		s: Box<dyn Subsystem<T>>,
```

Comment:
I don't really like the idea of subsystems being able to spawn other subsystems, whats the reasoning for this?

Reply:
This was one of the divergences from the implementor's guide we talked about on Riot before and I believe it is going to be changed